In [3]:
import pandas as pd
import numpy as np

trn = pd.read_csv('input/train_ver2.csv')

C:\Users\taeji\anaconda3\envs\kaggle_venv\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (5,8,11,15) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


데이터의 크기 / 첫 5줄 확인

- 총 13,647,309개의 고객 데이터
- 고객마다 48개의 변수가 존재

In [4]:
# trn.shape
trn.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0  2015-01-28   1375586            N              ES    H   35  2015-01-12   
1  2015-01-28   1050611            N              ES    V   23  2012-08-10   
2  2015-01-28   1050612            N              ES    V   23  2012-08-10   
3  2015-01-28   1050613            N              ES    H   22  2012-08-10   
4  2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo antiguedad  indrel  ... ind_hip_fin_ult1 ind_plan_fin_ult1  \
0        0.0          6     1.0  ...                0                 0   
1        0.0         35     1.0  ...                0                 0   
2        0.0         35     1.0  ...                0                 0   
3        0.0         35     1.0  ...                0                 0   
4        0.0         35     1.0  ...                0                 0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1 ind_valo_fin_ult1  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0             0.0                0.0                0  
1                0             0.0                0.0                0  
2                0             0.0                0.0                0  
3                0             0.0                0.0                0  
4                0             0.0                0.0                0  

[5 rows x 48 columns]

모든 변수에 대하여 미리 보기

In [8]:
for col in trn.columns:
    print('{}\n'.format(trn[col].head()))

0    2015-01-28
1    2015-01-28
2    2015-01-28
3    2015-01-28
4    2015-01-28
Name: fecha_dato, dtype: object

0    1375586
1    1050611
2    1050612
3    1050613
4    1050614
Name: ncodpers, dtype: int64

0    N
1    N
2    N
3    N
4    N
Name: ind_empleado, dtype: object

0    ES
1    ES
2    ES
3    ES
4    ES
Name: pais_residencia, dtype: object

0    H
1    V
2    V
3    H
4    V
Name: sexo, dtype: object

0     35
1     23
2     23
3     22
4     23
Name: age, dtype: object

0    2015-01-12
1    2012-08-10
2    2012-08-10
3    2012-08-10
4    2012-08-10
Name: fecha_alta, dtype: object

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: ind_nuevo, dtype: float64

0          6
1         35
2         35
3         35
4         35
Name: antiguedad, dtype: object

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: indrel, dtype: float64

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: ult_fec_cli_1t, dtype: object

0    1
1    1
2    1
3    1
4    1
Name: indrel_1mes, dtype: obje

**주의할 점**

- fecha_dato : 날짜 전용 데이터 타입인 datetime이 아닌 object이다.
- age        : 역시 데이터 타입이 object로 전처리 과정에서 int 타입으로 변환이 필요하다.
- renta      : 가구 총 수입을 나타냄. 5번째 열에서 NaN이 보임. 전처리 과정에서 이와 같은 결측값에 대한 변환도 필요하다.

### TIP 2-2 탐색적 데이터 분석을 통해 우리가 얻고자 하는 것?

> 새로 접하는 데이터에 대한 첫 분석 과정에서 데이터의 모든 것을 이해하려고 할 필요 없음!
랜덤하게 일부 행을 눈으로 살펴보며 단계적으로 데이터에 익숙해지려고 해보자
데이터에 대한 간단한 시각화도 큰 도움이 된다!
>
> - 아, 이번 경진대회 데이터는 이렇게 생겼구나
> - 이러한 변수들이 존재하는구나
> - data type을 보니, 이 변수는 전처리를 수행해야겠다
>
>정도의 느낌을 가져보자

In [9]:
trn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 48 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   fecha_dato             object 
 1   ncodpers               int64  
 2   ind_empleado           object 
 3   pais_residencia        object 
 4   sexo                   object 
 5   age                    object 
 6   fecha_alta             object 
 7   ind_nuevo              float64
 8   antiguedad             object 
 9   indrel                 float64
 10  ult_fec_cli_1t         object 
 11  indrel_1mes            object 
 12  tiprel_1mes            object 
 13  indresi                object 
 14  indext                 object 
 15  conyuemp               object 
 16  canal_entrada          object 
 17  indfall                object 
 18  tipodom                float64
 19  cod_prov               float64
 20  nomprov                object 
 21  ind_actividad_cliente  float64
 22  renta           

### 데이터 크기

앞서 .shape 함수를 통해 확인 했었다.

.info() 결과의 2, 3번째 줄 (RangeIndex, Data columns) 값으로 총 고객 데이터 수와 변수의 개수를 알 수 있다

### 변수

이어지는 48줄은 변수명과 해당 변수의 데이터 타입을 보여준다.

변수명이 스페인어로 구성되어 있는 것을 확인할 수 있다.

총 24개의 고객 관련 변수, 24개의 금융 제품 변수로 구성되어 있다.

### 데이터 타입

8개의 64bit float, 23개의 64bit int, 17개의 object 타입이 있다. 

머신러닝 모델 학습을 위해서는 훈련 데이터의 데이터 타입이 모두 int 혹은 float이어야 한다!

그러므로 전처리 과정에서 object 타입과 같은 변수를 적절하게 변환해서 모델 학습을 진행해야 한다.

### 메모리

.info() 를 통해 훈련 데이터의 메모리량도 확인 가능하다.

train_ver2.csv 파일은 2.2GB의 용량이지만, pandas를 통해 읽어오면 총 4.9GB의 메모리를 사용하게 된다.

불필요한 변수명 제거, 데이터 타입 변경 등을 통해 메모리를 효율적으로 사용할 수 있다.

---

### 수치형 / 범주형 변수

24개의 고객 관련 변수에 대해서 자세히 살펴보자.


#### 수치형

수치형 (int64, float64) 타입을 갖는 고객 변수를 "num_cols"로 추출하고 .describe()를 통해 간단한 요약 통계를 확인한다.

In [7]:
num_cols = [col for col in trn.columns[:24] if trn[col].dtype in ['int64', 'float64']]

trn[num_cols].describe()

ncodpers     ind_nuevo        indrel     tipodom      cod_prov  \
count  1.364731e+07  1.361958e+07  1.361958e+07  13619574.0  1.355372e+07   
mean   8.349042e+05  5.956184e-02  1.178399e+00         1.0  2.657147e+01   
std    4.315650e+05  2.366733e-01  4.177469e+00         0.0  1.278402e+01   
min    1.588900e+04  0.000000e+00  1.000000e+00         1.0  1.000000e+00   
25%    4.528130e+05  0.000000e+00  1.000000e+00         1.0  1.500000e+01   
50%    9.318930e+05  0.000000e+00  1.000000e+00         1.0  2.800000e+01   
75%    1.199286e+06  0.000000e+00  1.000000e+00         1.0  3.500000e+01   
max    1.553689e+06  1.000000e+00  9.900000e+01         1.0  5.200000e+01   

       ind_actividad_cliente         renta  
count           1.361958e+07  1.085293e+07  
mean            4.578105e-01  1.342543e+05  
std             4.982169e-01  2.306202e+05  
min             0.000000e+00  1.202730e+03  
25%             0.000000e+00  6.871098e+04  
50%             0.000000e+00  1.018500e+05  
75%             1.000000e+00  1.559560e+05  
max             1.000000e+00  2.889440e+07

24개 변수 중, 7개가 수치형 변수임을 확인 할 수 있다.

**.describe()** 함수는 pandas 데이터 프레임의 기초 통계를 보여준다.

min, max 값 사이의 precentile 값을 확인하는 것만으로도 해당 변수에 대한 이해를 높일 수 있다.

|번호|변수 이름|설명|비고|
|---|:---|:---|:---|
|1|nocodpers|고객 고유 식별 번호|(min) 15,889 ~ (max) 1,553,689 |
|2|ind_nuevo|신규 고객 지표|75%의 값이 0, 나머지(15%)가 값이 1|
|3|inderel|고객 등급 변수|75%의 값이 0, 나머지(15%)가 값이 99|
|4|tipodom|주소 유형 변수|모든 값이 1 (모든 값이 상수일 경우 변수로서 식별력을 가질 수 없음. 즉, 학습에 도움이 되지 않는 변수)|
|5|cod_prov|지방 코드 변수|(min) 1 ~ (max) 52의 값을 가지며, 수치형이지만 범주형 변수로서 의미를 가짐|
|6|ind_actividad_cliente|활발성 지표|50% 값이 0, 나머지(50%)가 값 1|
|7|renta|가구 총 수입|(min) 1,202.73 ~ (max) 28,894,400의 값|

#### 범주형

이번에는 (object) 데이터 타입을 갖는 범주형 변수를 "cat_cols"로 추출하고 간단한 요약 통계를 확인한다.

In [9]:
cat_cols = [col for col in trn.columns[:24] if trn[col].dtype in ['O']]

trn[cat_cols].describe()

fecha_dato ind_empleado pais_residencia      sexo       age  \
count     13647309     13619575        13619575  13619505  13647309   
unique          17            5             118         2       235   
top     2016-05-28            N              ES         V        23   
freq        931453     13610977        13553710   7424252    542682   

        fecha_alta  antiguedad ult_fec_cli_1t  indrel_1mes tiprel_1mes  \
count     13619575    13647309          24793   13497528.0    13497528   
unique        6756         507            223         13.0           5   
top     2014-07-28           0     2015-12-24          1.0           I   
freq         57389      134335            763    7277607.0     7304875   

         indresi    indext conyuemp canal_entrada   indfall   nomprov  \
count   13619575  13619575     1808      13461183  13619575  13553718   
unique         2         2        2           162         2        52   
top            S         N        N           KHE         N    MADRID   
freq    13553711  12974839     1791       4055270  13584813   4409600   

                 segmento  
count            13457941  
unique                  3  
top     02 - PARTICULARES  
freq              7960220

24개 변수 중, 17개가 범주형 변수이다.

수치형 변수에 대한 결과값과 조금 다른 것을 볼 수 있다. 각 행은 다음과 같은 의미가 있다.

|번호|결과행|설명|비고|
|-----|:---|:------|:---|
|1| count | 해당 변수의 유요한 데이터 개수 의미 | 'ult_fec_cil_1t'의 count가 24,793 밖에 확인되지 않으므로 결측값이 대부분임을 확인|
|2| unique | 해당 범주형 변수의 고유값 개수를 의미 | 성별 변수인 'sexo'에는 고유값이 2개임을 확인 |
|3| top | 가장 빈도가 높은 데이터 표시 | 나이 변수 'age'에서 최빈 데이터는 23세임을 확인 |
|4| freq | 최빈 데이터의 빈도수를 의미 | 총 데이터 수(count) 대비 최빈값(top)이 어느 정도인지에 따라 분포를 가늠할 수 있음.'ind_empleado'의 5개 고유값 중 가장 빈도가 높은 'N' 데이터가 전체의 99.9% 가량을 차지하며 데이터가 매우 편중되어 있음을 확인|

**주의해야할 점**

나이를 의미하는 'age', 은행 누적 거래 기간을 나타내는 'antiguedad'가 변수가 수치형이 아닌 범주형으로 분류되어 있다.

전처리 과정에서 수치형으로 변환해야 한다.

.
.

범주형 변수의 고유값을 직접 눈으로 확인해보자.

In [12]:
for col in cat_cols:
    uniq = np.unique(trn[col].astype(str))
    print('-' * 50)
    print(f'# col {col}, n_uniq {len(uniq)}, uniq{uniq}')

--------------------------------------------------
# col fecha_dato, n_uniq 17, uniq['2015-01-28' '2015-02-28' '2015-03-28' '2015-04-28' '2015-05-28'
 '2015-06-28' '2015-07-28' '2015-08-28' '2015-09-28' '2015-10-28'
 '2015-11-28' '2015-12-28' '2016-01-28' '2016-02-28' '2016-03-28'
 '2016-04-28' '2016-05-28']
--------------------------------------------------
# col ind_empleado, n_uniq 6, uniq['A' 'B' 'F' 'N' 'S' 'nan']
--------------------------------------------------
# col pais_residencia, n_uniq 119, uniq['AD' 'AE' 'AL' 'AO' 'AR' 'AT' 'AU' 'BA' 'BE' 'BG' 'BM' 'BO' 'BR' 'BY'
 'BZ' 'CA' 'CD' 'CF' 'CG' 'CH' 'CI' 'CL' 'CM' 'CN' 'CO' 'CR' 'CU' 'CZ'
 'DE' 'DJ' 'DK' 'DO' 'DZ' 'EC' 'EE' 'EG' 'ES' 'ET' 'FI' 'FR' 'GA' 'GB'
 'GE' 'GH' 'GI' 'GM' 'GN' 'GQ' 'GR' 'GT' 'GW' 'HK' 'HN' 'HR' 'HU' 'IE'
 'IL' 'IN' 'IS' 'IT' 'JM' 'JP' 'KE' 'KH' 'KR' 'KW' 'KZ' 'LB' 'LT' 'LU'
 'LV' 'LY' 'MA' 'MD' 'MK' 'ML' 'MM' 'MR' 'MT' 'MX' 'MZ' 'NG' 'NI' 'NL'
 'NO' 'NZ' 'OM' 'PA' 'PE' 'PH' 'PK' 'PL' 'PR' 'PT' 'PY' 'QA' 

--------------------------------------------------
# col indrel_1mes, n_uniq 10, uniq['1' '1.0' '2' '2.0' '3' '3.0' '4' '4.0' 'P' 'nan']
--------------------------------------------------
# col tiprel_1mes, n_uniq 6, uniq['A' 'I' 'N' 'P' 'R' 'nan']
--------------------------------------------------
# col indresi, n_uniq 3, uniq['N' 'S' 'nan']
--------------------------------------------------
# col indext, n_uniq 3, uniq['N' 'S' 'nan']
--------------------------------------------------
# col conyuemp, n_uniq 3, uniq['N' 'S' 'nan']
--------------------------------------------------
# col canal_entrada, n_uniq 163, uniq['004' '007' '013' '025' 'K00' 'KAA' 'KAB' 'KAC' 'KAD' 'KAE' 'KAF' 'KAG'
 'KAH' 'KAI' 'KAJ' 'KAK' 'KAL' 'KAM' 'KAN' 'KAO' 'KAP' 'KAQ' 'KAR' 'KAS'
 'KAT' 'KAU' 'KAV' 'KAW' 'KAY' 'KAZ' 'KBB' 'KBD' 'KBE' 'KBF' 'KBG' 'KBH'
 'KBJ' 'KBL' 'KBM' 'KBN' 'KBO' 'KBP' 'KBQ' 'KBR' 'KBS' 'KBU' 'KBV' 'KBW'
 'KBX' 'KBY' 'KBZ' 'KCA' 'KCB' 'KCC' 'KCD' 'KCE' 'KCF' 'KCG' 'KCH' 'KCI'
 'KCJ' 'KC